### PR TITLE
[test_runner] Unify wait routines

### PR DESCRIPTION
**Pull request by marcoguerri**
The pipeline code was affected by some duplication in the shutdown
control path. This patch removes the duplication, and introduces
only two wait routines in the shutdown control path: waitTargets
and waitControl.

### CODE REVIEW

File: cmds/contest/hosts02.csv [hunk 1]
```diff
@@ -1,4 +1 @@
 compute1234,172.16.5.6
-storage2345,10.10.10.123
-machinelearning1234,192.168.123.231
-uselesshost10,10.0.0.10
```
Comment (xaionaro) [inline cmds/contest/hosts02.csv:4]: **out of curiosity:**

Why have you removed this lines? :)

File: cmds/clients/contestcli-http/start.json [hunk 2]
```diff
@@ -20,23 +20,6 @@
...
-                "HostPrefixes": [
-                ]
-            },
-            "TargetManagerReleaseParameters": {
-            },
-            "TestFetcherName": "URI",
-            "TestFetcherFetchParameters": {
-                "TestName": "RackProvisioning",
-                "URI": "test_samples/randecho.json"
-            }
```
Comment (xaionaro) [inline cmds/clients/contestcli-http/start.json:39]: **out of curiosity:**

Why have you removed this lines? :)

File: cmds/contest/test_samples/randecho.json [hunk 3]
```diff
@@ -1,26 +1,28 @@
...
-            "label": "thethirdrandecho",
+            "name": "randecho",
+            "label": "thefirstrandecho_cleanup",
             "parameters": {
-                "text": ["some very random text"]
+                "text": ["some very random text first_cleanup"]
             }
         }
-    ]
+   ]
```
Comment (xaionaro) [inline cmds/contest/test_samples/randecho.json:26]: **extreme nitpicking:**

An extra space.

File: pkg/runner/test_runner_pipeline.go [hunk 4]
```diff
@@ -198,174 +207,161 @@ func (p *pipeline) runStep(cancel, pause <-chan struct{}, jobID types.JobID, run
 	}
 }
 
-// waitTargets reads results coming from results channels until all Targets
-// have completed or an error occurs. If all Targets complete successfully, it checks
-// whether Steps and routing blocks have completed as well. If not, returns an
-// error. Termination is signalled via terminate channel.
-func (p *pipeline) waitTargets(terminate <-chan struct{}, completedCh chan<- *target.Target) error {
+// emitStepEvent emits a failure event if a step fail
+func (p *pipeline) emitStepEvent(result *stepResult) error {
```
Comment (xaionaro) [inline pkg/runner/test_runner_pipeline.go:211]: I interpret this name differently from what it does: I interpret that it sends an event on either success or failure, while actually it sends an event only on failure. BTW, why we do not send successes? To reduce amount of rows in DB?

File: pkg/runner/test_runner_pipeline.go [hunk 5]
```diff
@@ -198,174 +207,161 @@ func (p *pipeline) runStep(cancel, pause <-chan struct{}, jobID types.JobID, run
...
+	return ev.Emit(errEv)
+}
 
-	var (
-		err                  error
-		completedTarget      *target.Target
-		completedTargetError error
-	)
+// waitControl reads results coming from result channels (for steps and routing blocks)
+// until a timeout occurrs. The error handling is different depending on whether
```
Comment (xaionaro) [inline pkg/runner/test_runner_pipeline.go:239]: **nitpicking:**

A typo: `occurrs` → `occurs`.

**Comment by marcoguerri** [comment]
Why was this...re-opened?

**Pull request closed without merging**